### ISSUE TITLE
Adding an input to the middle of a block's prototype should be smart about references to the block

### ISSUE THREAD

**Issue by brianharvey** [jmoenig/Snap--Build-Your-Own-Blocks#1054]
When you add an input to the prototype of a custom block, in the block editor, all the instances of that block get an extra input slot.  But the new slot always comes at the end -- that is, the existing input expressions remain associated with the first N-1 input slots.

This is the right thing if the new formal parameter is at the end, but not if the new parameter is in the middle someplace.  In that case, the existing inputs should move right so as to remain bound to the same formal parameters as before.

The discussion in #1049 reminded me of this.

**Comment by jmoenig** [comment]
Variables are late-bound dynamically at runtime, they now can already be polymorphic. See, your suggestions often boil down to static links between individual syntax elements and their meaning. I can see how a naive novice might expect such static links in a visual programming language, but Snap! is different, right? What if your project uses CALL to invoke a ringified block and passes some inputs, or even an input list. Now if you change the definition of the ringified block you could argue that the IDE should be smart enough to detect every CALL, RUN and LAUNCH. But even if you *just* want to be smart about inputs in blocks-instances you'd have to come up with a way to apply a certain "insert" operation to each instance, whereas we currently simply refresh the block. I'm not saying it cannot be done, but - again - I can see useful steps in that direction that would also help us with other issues, and *tools* such as being able to browse all callers would definitely help.

**Comment by brianharvey** [comment]
I started thinking about this partly because of a conversation with Dan.  He tried to add a STEP input to the FOR block's definition, and as a result, all of the scripts inside C-slots of FOR instances disappeared altogether, since a script can't go into a numeric input slot.  In that situation, applying some tool after the fact is too late to help.  And the right kind of tool -- namely one that does something to every instance of something -- is hard to express visually.  It's the sort of thing you do with a one-line shell script or Perl script.

> I can see how a naive novice might expect such static links in a visual programming language, but Snap! is different, right?

I don't know; all else being equal, I think Snap<i>!</i> *should* do what that naive user expects.  All else is rarely equal, of course.  I'm not suggesting that <(x) = <(7) or (11)>> should do what the user expects, although we should recognize that pattern and give a very specific error message with a suggested correction.  But in this case of adding a formal parameter, I think the naive user's expectation is *right*.

**Comment by jmoenig** [comment]
Yeah, right. So, this is another of these things that make me all mad, because, look, I've made it so much more flexible and extensible to edit blocks in Snap, to add, rename and remove parameters and to change categories etc., and then people are filing *issues* because it's not *enough*.

Seriously, have you even looked at Scratch? Have you tried the same in Tynker? Did you try inserting an input or a label part anywhere into an existing block (hint: it's possible in both environments, Tynker and Scratch, but did you notice what you have to do accomplish this?), did you then observe what happens to existing block instances and how it arranges existing inputs, either embedded reporters or user-edited contents? Did you try any of that in any other language?

This is exactly why I don't even want to talk about introducing new features or giving in to user feature requests, because - sure enough - what *will* happen is that people will complain about it even more, because *now* they want to to something entirely different and somehow feel that *that's* what Snap "should" do:

This has been happening time and again:

* Someone asks for "high-resolution" pictures of scripts, so they can be printed on posters. I add a feature that lets you zoom blocks arbitrarily. Next thing I see is everybody is *only* using the blocks zooming feature and complains about everything else *not* being zoomable
* I add the capability to change block labels, so in Snap - unlike in Scratch - you don't have to rip out a block from its script, throw it away and then insert the one you want, but instead you can re-label it in place. And I let you do the same with variables, even offering a menu that lets you choose from all reachable variable names within the context scope of that block. Sure enough next thing that happens is people complaining about missing goddam *refactoring* support in Snap

The lesson I take away from this is *not* to add features because it'll make users unhappy and complain, it'll give them the impression that Snap is somehow defective.

**Comment by cycomachead** [comment]
I have a longer post here, that I'm working on, but gosh -- feature requests are expected and good because it means students and others are using Snap<em>!</em>. 

A lot of these things are because we are fighting multiple issues with using a visual programming language. We are trying to tell students that they don't need to "move on" from Snap<em>!</em>, and they can build cooler/better/more-X projects in Snap<em>!</em>. We _can_ but the walls that students hit aren't conceptual or because of a limitation of power. The vast majority are because in any other non-visual environment, they have additional, near infinite tools they can use and learn at their own pace to solve problems. Snap<em>!</em> is both a language **AND** an IDE. I honestly believe that (aside from minor browser quibbles) the vast majority of problems students run into are with the IDE side of things. 

And yeah, Tynker and Scratch, and so many other things are nowhere near what we could do in Snap<em>!</em>. No one so far is debating that.

**Comment by cycomachead** [comment]
I _just_ lost code in this manner, but by deleting inputs from a function. The code literally disappears. That is definitely not expected and would be really frustrating if I weren't doing a test project. 

(But I do have a deadline I need to finish (using Snap<em>!</em>), so it _is_ still pretty annoying.)

**Comment by jmoenig** [comment]
Wait! Snap *is* preserving complex inputs, i.e. reporters, by placing them outside of the changed blocks. Try doing the same in Scratch!

What Snap currently does not preserve is scripts inside C-shaped slots if you change the C-shaped slot to something else, or if you delete it altogether.

**Comment by cycomachead** [comment]
> What Snap currently does not preserve is scripts inside C-shaped slots if you change the C-shaped slot to something else, or if you delete it altogether.

Or, if you add/remove an input. This just lost me 20 mins of work. And I have a deadline very soon....

**Comment by bb010g** [comment]
Could we pop out the C-block lambda into an actual ring?

**Comment by brianharvey** [comment]
> So, this is another of these things that make me all mad, because, look, I've made it so much more flexible and extensible to edit blocks in Snap, to add, rename and remove parameters and to change categories etc., and then people are filing *issues* because it's not *enough.*

Yes.  That's what happens when you have users; they expect, so to speak, the closure of the features you give them.  They find all the edge cases.  They start using the features in ways you didn't anticipate.

Users are greedy and unthankful.  (Actually, that's not true, I hear all the time from teachers who get what an amazing thing Snap<i>!</i> is and are grateful for it.  But it doesn't stop them filing bug reports.  :-))  And they think this is like a Microsoft product with thousands of QA people just waiting to solve problems.

I mean, for example, you and I talked forever about grey rings, and it wasn't until you implemented them and I could actually use them that I realized that dragging a ring with formal parameters into a procedure-type input slot should keep the formal parameters when absorbing the ring.  I keep being bitten by that one!  Ungrateful me.

So, what software developers do is hang out with other software developers and complain to each other about what jerks the users are, and then do what the users want -- unless, again, the users are actually wrong, as in the <7 or 11> case.  But this time the users are right; when a user adds a parameter in the middle of the prototype, you know which is the new one, and you can preserve the bindings of the old parameters to their actual input expressions in instances of the block.

You get to complain to users if they flag every issue as urgent.  (I didn't flag this one as urgent!)  You get to set priorities for fixing things, and you get to set the balance between fixing old things and creating new things, which is the fun part.  But I don't think you get to complain because people file issues; you should expect that.
